File: Leetcode/stack_queue/150_EvaluateReversePolishNotation/Solution.py
class Solution(object):
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        ops = { "+": operator.add, "-": operator.sub, "*": operator.mul, "/":operator.floordiv}
        stack = []
        for token in tokens:
            if token not in ops:
                stack.append(int(token))
            elif token=="/":
                sign = -1 if stack[-2]*stack[-1] <0 else 1
                stack.append( sign*ops[token]( abs(stack.pop(-2)), abs(stack.pop()) ) )
            else:
                stack.append( round(ops[token]( stack.pop(-2), stack.pop() )) )
        return int(stack[-1])

File: Leetcode/stack_queue/150_EvaluateReversePolishNotation/Solution.java
class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        String[] arr = {"+","-","*","/"};
        Set<String> ops = new HashSet<>(Arrays.asList(arr));
        for(String token: tokens){
            if(!ops.contains(token)){
                stack.push(Integer.parseInt(token));
            }else{
                int num2 = stack.pop(), num1 = stack.pop();
                if(token.equals("+")){
                    stack.push(num1+num2);
                }else if(token.equals("-")){
                    stack.push(num1-num2);
                }else if(token.equals("*")){
                    stack.push(num1*num2);
                }else{
                    stack.push(num1/num2);
                }
            }
        }
        return stack.peek();
    }
}
